### ISSUE TITLE
connmand Does Not Seem to Failover on Anything But Local Link Failures; Implement Non-local Link Failure Failover

### ISSUE THREAD

**Issue by gerickson** [Nuovations/connman#6]
I spent most of Friday (2023-10-13) doing integration testing among `connmand`, inflicting various communications failures and observing how the event chain responds. I learned that `connmand` appears to only be responsive to local link failures:

* That is, with Ethernet, only a first hop link loss or cable pull will cause a failover to Wi-Fi. Likewise, with Wi-Fi, only a complete loss of association will cause a failover to Cellular.
    * So, for example, removing an Ethernet link several hops away from the controller or from the Wi-Fi access point, compromising DNS, etc. do not cause default service failovers (though I can clearly see the HTTP liveness probes failing with 4xx status—so the right signals are present to induce a failover).
* At least for embedded use cases, I don’t believe this is sufficient.
* I would propose some changes to `connmand` upstream that:
    * After some configurable number of app-layer HTTP liveness probes failing, `connmand` declares a technology / service no longer “online”
    * In response, `connmand` moves the default route to the next available technology / service (for example, Wi-Fi or Cellular).
    * `connmand` would then go into TBD (Fibonacci?) back off for liveness on the “no longer online” link(s) and would promote it / them only when some configurable number of app-layer HTTP liveness probes succeeds.

This would appear to continue the work started by @evautrin with his `EnableOnlineToReadyTransition` work and patches.

**Comment by gerickson** [comment]
Addressing #5 will "paper over it"; however, fundamentally, there is the same high- vs. low-priority routing metric at play here for online check failover. For the online check / WiSPr, host routes for the IPv4 or IPv6 URL host are installed and removed.

### Installed

```c
static bool wispr_route_request(const char *address, int ai_family,
        int if_index, gpointer user_data)
{
    int result = -1;
    struct connman_wispr_portal_context *wp_context = user_data;
    const char *gateway;
    struct wispr_route *route;

    gateway = __connman_ipconfig_get_gateway_from_index(if_index,
        wp_context->type);

    DBG("address %s if %d gw %s", address, if_index, gateway);

    if (!gateway)
        return false;

    route = g_try_new0(struct wispr_route, 1);
    if (route == 0) {
        DBG("could not create struct");
        return false;
    }

    switch (wp_context->type) {
    case CONNMAN_IPCONFIG_TYPE_IPV4:
        result = connman_inet_add_host_route(if_index, address,
                gateway);
        break;
    case CONNMAN_IPCONFIG_TYPE_IPV6:
        result = connman_inet_add_ipv6_host_route(if_index, address,
                gateway);
        break;
    case CONNMAN_IPCONFIG_TYPE_UNKNOWN:
    case CONNMAN_IPCONFIG_TYPE_ALL:
        break;
    }

    if (result < 0) {
        g_free(route);
        return false;
    }

    route->address = g_strdup(address);
    route->if_index = if_index;
    wp_context->route_list = g_slist_prepend(wp_context->route_list, route);

    return true;
}
```

### Removed

```c
static void free_wispr_routes(struct connman_wispr_portal_context *wp_context)
{
    while (wp_context->route_list) {
        struct wispr_route *route = wp_context->route_list->data;

        DBG("free route to %s if %d type %d", route->address,
                route->if_index, wp_context->type);

        switch (wp_context->type) {
        case CONNMAN_IPCONFIG_TYPE_IPV4:
            connman_inet_del_host_route(route->if_index,
                    route->address);
            break;
        case CONNMAN_IPCONFIG_TYPE_IPV6:
            connman_inet_del_ipv6_host_route(route->if_index,
                    route->address);
            break;
        case CONNMAN_IPCONFIG_TYPE_UNKNOWN:
        case CONNMAN_IPCONFIG_TYPE_ALL:
            break;
        }

        g_free(route->address);
        route->address = NULL;

        g_free(route);
        wp_context->route_list->data = NULL;

        wp_context->route_list =
            g_slist_delete_link(wp_context->route_list,
                    wp_context->route_list);
    }
}
```

However, note that there is not a metric parameter in `connman_inet_{add,del}{_ipv6,}_host_route`. So, in a failover scenario, there may be host routes installed for one or more of Ethernet, Wi-Fi, and Cellular as each tries to recover online connectivity. However, since the metric for all is implicitly zero, one and only one network interface index will get the host route and all the others will silently fail (because `connman_inet_{add,del}{_ipv6,}_host_route` silently swallows `-EEXIST`).

In the recent set of patches to #5, I'd commented in `compute_low_priority_metric`:

```c
static uint32_t compute_low_priority_metric(const struct gateway_data *data)
{
    static const uint32_t metric_base = UINT32_MAX;
    static const uint32_t metric_ceiling = (1 << 20);
    static const uint32_t metric_index_step = (1 << 10);

    /*
     * The algorithm uses the network interface index since it is
     * assumed to be stable for the uptime of the network interface
     * and, consequently, the potential maximum lifetime of the route.
     *
     * The algorithm establishes UINT32_MAX as the metric base (the
     * lowest possible priority) and a somewhat-arbitrary 2^20 as the
     * ceiling (to keep metrics out of a range that might be used by
     * other applications). The metric is then adjusted in increments
     * of 1,024 (2^10) from the base, but less than the ceiling, by
     * multiplying the increment by the network interface index. This
     * is easy and simple to compute and is invariant on service
     * order.
     *
     * In the fullness of time, the "rule of least astonishment" for
     * Connection Manager might be that low priority metrics follow
     * the service order with the default service always having metric
     * zero (0) and lowest priority metric assigned to the lowest
     * priority service, etc. Achieving this would require 1) caching
     * the computed metric in the gateway data since services may
     * re-sort by the time we are asked to recompute high- and
     * low-priority routes and we need a stable and matching metric to
     * successfully delete a previously-created route and 2) having
     * access to an API (such as
     * '__connman_service_get_order(data->service)') that exposes a
     * strictly-in/decreasing service order with no duplicates. Today,
     * there is no such API nor is there such a durable service order
     * meeting that mathematical requirement.
     */
    return MAX(metric_ceiling,
                metric_base -
                (data->index * metric_index_step));
}
```

However, because this routing metric is needed across multiple modules (right now, both wispr.c and connection.c) the right approach given this revelation might be an API:

```
__connman_service_get_route_metric(const struct connman_service *service);
```

where the default service returns zero (`0`) and the rest return a metric reflective of their service order, between some ceiling (say, `2^20`) and some floor (say, `UINT32_MAX`).

This could then be applied easily to both gateway network and on-link network routes as well, making routing determinant based on service order, not kernel preference, for all routes.

**Comment by gerickson** [comment]
At this point, with `OnlineCheckMode=none`, I have Ethernet + Wi-Fi (with Ethernet as preferred) and I can repeatedly connect/disconnect the Ethernet link directly at the device-under-test (DUT) and fail over to Wi-Fi and back to Ethernet with each service assuming the correct high- (metric 0) and low-priority (metric > 0) routes, with Wi-Fi still able to ping `dns.google.com` over IPv4 when Ethernet is the default service.

The [attached log](https://github.com/Nuovations/connman/files/13721326/connman-github-issue-6-log-annotated-output-20231219T153722.txt) demonstrates that success.

Next up, same scenario but with `OnlineCheckMode=one-shot`.

**Comment by gerickson** [comment]
At this point, with `OnlineCheckMode=one-shot`, I have Ethernet + Wi-Fi (with Ethernet as preferred) and I can repeatedly connect/disconnect the Ethernet link directly at the device-under-test (DUT) and fail over to Wi-Fi and back to Ethernet with each service assuming the correct high- (metric 0) and low-priority (metric > 0) routes, with Wi-Fi still able to ping `dns.google.com` over IPv4 when Ethernet is the default service.

The [attached log](https://github.com/Nuovations/connman/files/13721571/connman-github-issue-6-log-annotated-output-20231219T164738.txt) demonstrates that success.

Next up, same scenario but with `OnlineCheckMode=continuous`.

**Comment by gerickson** [comment]
At this point, with `OnlineCheckMode= continuous`, I have Ethernet + Wi-Fi (with Ethernet as preferred) and I can repeatedly connect/disconnect the Ethernet link directly at the device-under-test (DUT) and fail over to Wi-Fi and back to Ethernet with each service assuming the correct high- (metric 0) and low-priority (metric > 0) routes, with Wi-Fi still able to ping `dns.google.com` over IPv4 when Ethernet is the default service.

**Comment by gerickson** [comment]
Closed with #121.
